### PR TITLE
Added code review pattern

### LINKED ISSUE #532: Code review pattern
Our engineering patterns are example solutions for addressing problems, meeting standards and living our engineering principles. They should make it easier for engineers to do their jobs. We don't have a standard way of writing patterns as we like to be a bit more flexible with their structure. Please provide us with some information on what you are proposing below

**What is the pattern you are suggesting?**
It will be used to show what code review best practices are.

**What would be the benefit to the Home Office of following this pattern**
Improve the quality of code and general developer behaviour.

**How do people implement this pattern**
When working through a code review

- [X] I have looked through our existing patterns and think this is not covered elsewhere
- [X] I have checked through the open issues on the repository and this pattern has not already been suggested

### PR DESCRIPTION
**Pull request by aaronrussellHO**
Is this pull request a content or a code change? (Please fill in the relevant section and delete the other)

# Code change
I can confirm:
## Accessibility considerations
- [X] Please review the [accessibility checks for layout changes](https://github.com/UKHomeOffice/engineering-guidance-and-standards/blob/main/technical-docs/accessibility/layout-checks.md).

- [x] This change will not change layouts, page structures or anything else that might impact accessibility

or
- [ ] This change might impact accessibility, automated aXe tests cover the impact

or
- [ ] This change might impact accessibility and is not covered by automated aXe tests - manual testing has been performed
(If the change might impact accessibility then please add some further information here)

## Commit signing

- [x] All commits are signed with a key that can be verified by GitHub. [GitHub guidance on signing commits](https://docs.github.com/en/enterprise-cloud@latest/authentication/managing-commit-signature-verification/signing-commits)

# Content change 
- [x] Please review the [accessibility checks for content changes](https://github.com/UKHomeOffice/engineering-guidance-and-standards/blob/main/technical-docs/accessibility/content-checks.md).

I can confirm:
- [x] Content does not include any code or configuration changes (excluding frontmatter information)
- [x] Content meets the content standards
e.g. [Writing a principle](https://engineering.homeoffice.gov.uk/standards/writing-a-principle/) and [Writing a standard](https://engineering.homeoffice.gov.uk/standards/writing-a-standard/)
- [x] Content is suitable to open source, i.e.:
    - Content does not relate to unreleased gov policy
    - Content does not refer to anti-fraud mechanisms
    - Content does not include sensitive business logic
- [x] Last updated date for content is correct
- [x] All commits are signed with a key that can be verified by GitHub. [GitHub guidance on signing commits](https://docs.github.com/en/enterprise-cloud@latest/authentication/managing-commit-signature-verification/signing-commits)

### CODE REVIEW

File: docs/patterns/code-reviews.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+---
+
+## Considerations
+
+### Some of the things that should be checked for
+
+This is not an exaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
+
+#### Does it work as attended?
```
Comment (daniel-ac-martin) [inline docs/patterns/code-reviews.md:45]:
```suggestion
#### Does it work as intended?
```

**Comment by daniel-ac-martin** [comment]
For pull requests that include AI generated code, I think we should consider whether we need to mandate two independent code reviews. (That might be beyond the intended scope of this page though.)

**Comment by daniel-ac-martin** [comment]
Some more ideas:
- Does the code adequately defend against invalid inputs?
- Does the code correctly handle all error cases from the function it calls?
- Does it clean up the resources it is responsible for? (e.g. Closing file handlers.)
- Steal something good from the internet? - Some ideas:
  - https://snappify.com/blog/code-review-checklist-template
  - https://google.github.io/eng-practices/review/reviewer/looking-for.html
  - https://typoapp.io/blog/code-review-best-practices/

Those are just some very vague ideas + a quick bit of Google searching. I'm sure it would be possible to put things into more general principles / areas of consideration.

File: docs/patterns/code-reviews.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+Positive feedback can help build a positive culture within your team and it's always nice to hear positive feedback.
+
+### Have a pull request template
+
+A pull request template can help you create a check list of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
+
+### Support your code reviews with automation
+
+Automate as much as possible to help relive the code review burdon, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:35]:
```suggestion
Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.
```

File: docs/patterns/code-reviews.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+Time should be allocated as part of your sprint to review code, as it's important to keep high coding standards.
+
+### Code reviews should be constructive and friendly
+
+When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.
+
+### Point out the good things as well as bad
+
+Positive feedback can help build a positive culture within your team and it's always nice to hear positive feedback.
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:27]:
```suggestion
Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.
```

File: docs/patterns/code-reviews.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+### Have a pull request template
+
+A pull request template can help you create a check list of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:31]:
```suggestion
A pull request template can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
```

**Review by jeff-horton-ho-sas — COMMENTED** [review]
Typos/grammar

File: docs/patterns/code-reviews.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.
+
+### Point out the good things as well as bad
+
+Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.
+
+### Have a pull request template
+
+A pull request template can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:31]:
```suggestion
A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too. 
```

File: docs/patterns/code-reviews.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
+
+#### Do the automated tests pass?
+
+Does it pass automated testing? Does it pass linting checks?
+
+#### Do regular expessions make sense?
+
+Does a regex work as expected? Does it include all edge cases?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:83]: Does this work better as a definition/summary list?

```suggestion
Does it work as intended?
: Does the code look like it would work like the author intended? Does it work as users would expect?

Is it tested?
: Are the tests well designed? Is there enough coverage? Are the tests automated?

Is it easy to read?
: Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?

Is the software well designed and bug free?
: Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?

Could the code be used for something other than the intended purpose?
: Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?

Does any of the code duplicate an existing function?
: Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?

Is the code documented?
: Should there be more documentation around this code? Would you need more information before working on it?

Does it follow standards?
: Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?

Do the automated tests pass?
: Does it pass automated testing? Does it pass linting checks?

Do regular expessions make sense?
: Does a regex work as expected? Does it include all edge cases?
```

image

**Review by jeff-horton-ho-sas — APPROVED** [review]
A couple of suggestions, but neither blocks publishing

File: docs/patterns/code-reviews.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?
+
+#### Is the code documented?
+
+Should there be more documentation around this code? Would you need more information before working on it?
+
+#### Does it follow standards?
+
+Does it follow the teams standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:75]:
```suggestion
Does it follow the team's standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
```

File: docs/patterns/code-reviews.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+#### Do the automated tests pass?
+
+Does it pass automated testing? Does it pass linting checks?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:79]:
```suggestion
Does it pass automated testing? Does it pass linting checks? Have tests been added to validate new functionality?
```

File: docs/patterns/code-reviews.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+
+Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?
+
+#### Could the code be used for something other than the intended purpose?
+
+Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?
+
+#### Does any of the code duplicate an existing function?
+
+Is there any code in the codebase doing something similar? Can something be extended instead? Has the code been generated by AI and is not aware of shared functions/services etc?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:67]:
```suggestion
#### Do the changes fit well into the existing code?
Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and convetions of the team. Does it make good use of the application's framework?

Pay particular attention to this if the code may have generated by AI which is not aware of shared functions/services etc?
```

File: docs/patterns/code-reviews.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
+
+#### Is the software well designed and bug free?
+
+Is the code bug free? Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any obvious security flaws?
+
+#### Could the code be used for something other than the intended purpose?
+
+Does the code have any security flaws? Do error messages contain too much information? Is there validation on user input?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:63]:
```suggestion
#### Does the code protect against misuse?

Does the code have any security vulnerabilities, e.g. does it allow code injection? Do error messages contain too much information? Is user input validated?
```

File: docs/patterns/code-reviews.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+Does the code look like it would work like the author intended? Does it work as users would expect?
+
+#### Is it tested?
+
+Are the tests well designed? Is there enough coverage? Are the tests automated?
+
+#### Is it easy to read?
+
+Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md]:
```suggestion
#### Is it easy to read?

Can you clearly understand what the code is doing? Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
```

File: docs/patterns/code-reviews.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+
+Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.
+
+---
+
+## Considerations
+
+### Some of the things that should be checked for
+
+This is not an exaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:43]:
```suggestion
This is not an exhaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
```

File: docs/patterns/code-reviews.md [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+: Can you clearly understand what the code is doing? Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
+
+Is the software well designed?
+: Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any security vulnerabilities, performance issues, or other bugs?
+
+Does the code protect against misuse?
+: Does the code have any security vulnerabilities, e.g. does it allow code injection? Do error messages contain too much information? Is user input validated?
+
+Do the changes consider existing code?
+: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and convetions of the team. Does it make good use of the application's framework?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:61]:
```suggestion
: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team. Does it make good use of the application's framework?
```

File: docs/patterns/code-reviews.md [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+
+  Pay particular attention to this if the code may have generated by AI which is not aware of shared functions/services etc?
+
+Is the code documented?
+: Should there be more documentation around this code? Would you need more information before working on it?
+
+Does it follow standards?
+: Does it follow the team's standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
+
+Do regular expessions make sense?
```
Comment (jeff-horton-ho-sas) [inline docs/patterns/code-reviews.md:71]:
```suggestion
Do regular expressions make sense?
```

**Review by jeff-horton-ho-sas — COMMENTED** [review]
Typo regressions

File: docs/patterns/code-reviews.md [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+When reviewing code, you should focus on the code, not the author. Remember that every developer has different ideas and solutions to problems. Using reasons to explain why something needs changing is much more valuable than expressing your feelings.
+
+### Point out the good things as well as bad
+
+Positive feedback can help build a positive culture within your team, and it's always nice to hear positive feedback.
+
+### Have a pull request template
+
+A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code too.
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:31]:
```suggestion
A [pull request template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository) can help you create a checklist of what to do before/while creating a pull request, it can also be useful for the person reviewing the code.
```
Probably don't need both "also" + "too" in this last sentence.

File: docs/patterns/code-reviews.md [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+This is not an exhaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
+
+Does it work as intended?
+: Does the code look like it would work like the author intended? Does it work as users would expect?
+
+Is it tested?
+: Are the tests well designed? Is there enough coverage? Are the tests automated? Does it pass automated testing? Does it pass linting checks?
+
+Is it easy to read?
+: Can you clearly understand what the code is doing? Are variables, functions etc. names descriptive? Is the code too complex? Are comments useful and clear? Are the comments needed?
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:52]: Small wording suggestion

```suggestion
: Can you clearly understand what the code is doing? Do variables, functions and classes have descriptive names? Is the code too complex? Are comments useful and clear? Are the comments needed?
```

File: docs/patterns/code-reviews.md [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+Is the software well designed?
+: Is there good error handling? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any security vulnerabilities, performance issues, or other bugs?
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:55]: Security vulnerabilities are covered in the point below, so maybe don't need it in this one as well.

Also, may make sense to consider logging alongside error handling?

```suggestion
: Is there good error handling and logging? Does it fit with the rest of the code in the project? Have they thought about edge cases? Can you see any performance issues or other bugs?
```

File: docs/patterns/code-reviews.md [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+
+Does the code protect against misuse?
+: Does the code have any security vulnerabilities, e.g. does it allow code injection? Do error messages contain too much information? Is user input validated?
+
+Do the changes consider existing code?
+: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team. Does it make good use of the application's framework?
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:61]:
```suggestion
: Is there existing code doing something similar? Can a shared function be extended instead? Does it match the style and conventions of the team? Does it make good use of the application's framework?
```

File: docs/patterns/code-reviews.md [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+  Pay particular attention to this if the code may have generated by AI which is not aware of shared functions/services etc?
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:63]:
```suggestion
  Pay particular attention to this if the code may have been generated by AI which is not aware of shared functions/services etc.
```

File: docs/patterns/code-reviews.md [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+Is the code documented?
+: Should there be more documentation around this code? Would you need more information before working on it?
+
+Does it follow standards?
+: Does it follow the team's standards? Does it follow [SEGAS standards](/standards/)? Does it follow [service manual standards](https://www.gov.uk/service-manual)?
+
+Do regular expressions make sense?
+: Does a regex work as expected? Does it include all edge cases?
+
+---
```
Comment (PaulWrightHO) [inline docs/patterns/code-reviews.md:74]:
```suggestion
```
Not sure if it's needed but this section break makes it look like there's an empty row at the bottom of the table, similar to the bottom of https://engineering.homeoffice.gov.uk/patterns/docs-as-code/ 

Seems to render fine when removed locally - but again, not sure if it's needed for something else.

**Review by PaulWrightHO — APPROVED** [review]
Looks like some tests failed, but happy to approve the content

**Comment by jeff-horton-ho-sas** [comment]
> Looks like some tests failed, but happy to approve the content

Failed tests were due to gov.uk links timing out - passing now.

**Comment by aaronrussellHO** [comment]
> > Looks like some tests failed, but happy to approve the content
> 
> Failed tests were due to gov.uk links timing out - passing now.

Thanks for sorting!

File: docs/patterns/code-reviews.md [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+
+### Some of the things that should be checked for
+
+This is not an exhaustive list, just some recommendations, but your team should have their own standards, a pull request template can help with this too by creating a checklist of what to look for.
+
+Does it work as intended?
+: Does the code look like it would work like the author intended? Does it work as users would expect?
+
+Is it tested?
+: Are the tests well designed? Is there enough coverage? Are the tests automated? Does it pass automated testing? Does it pass linting checks?
```
Comment (RobJohnsonHO) [inline docs/patterns/code-reviews.md:49]: I think we could reformat this instead of a set of questions.  As a pattern we should suggest some good practices for code reviews.
Something like "Testing in code is important, make sure to check that current tests are passed and make suggestions where further automations would help"

**Review by RobJohnsonHO — COMMENTED** [review]
Overall I think this is a great start!
I'm not sure the format of questions in the considerations section support the pattern.
Maybe we could also add some information around the size of commits help the speed of a code review.  The more files changes the harder and longer it can take.
It could link back out to the principles for small changes.

File: docs/patterns/code-reviews.md [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+Is the software well designed?
+: Checking for things such as good error handling and logging can help developers debugging and provide a better user experience. Making sure the code fits with the rest of the project is also important. Finally checking if they have missed any edge cases or you can see any performance issues or bugs.
+
+Does the code protect against misuse?
+: Protecting against security vulnerabilities is crucial, checking for things such as allowing code injection, not validating user inputs or including Personally Identifiable Information (PII) in error messages will help to catch problems early.
+
+Do the changes consider existing code?
+: Reusability is important because it enhanced efficiency and minimises errors and leads to faster development and makes code easier to maintain. Frameworks, tools and standards within a team can also help a developer reuse functions they provide, make sure to check that code can't make use of these sources. Pay particular attention to this if the code may have been generated by AI which is not aware of shared functions/services etc.
```
Comment (RobJohnsonHO) [inline docs/patterns/code-reviews.md:61]: There is a typo here: 
Reusability is important because it **enhances** efficiency

File: docs/patterns/code-reviews.md [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+
+Automate as much as possible to help relieve the code review burden, things like testing and linters can be automated. Tools like ESLint and Prettier can help with this.
+
+---
+
+## Considerations
+
+### Some of the things that should be checked for
+
+This is not an exhaustive list, just some recommendations, but your team should have their own standards. A pull request template can help with this by creating a checklist of what to look for.
```
Comment (RobJohnsonHO) [inline docs/patterns/code-reviews.md:43]: This is not an exhaustive list, just some recommendations.  Your team should ...

File: docs/patterns/code-reviews.md [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+
+Does it work as intended?
+: Ensuring that code works as intended is a crucial part of the code review process because it verifies functionality, preventing bugs and errors from reaching production, which saves time and resources. It also ensures a smooth and reliable user experience, maintaining user trust and satisfaction.
+
+Is it tested?
+: Testing code is important, make sure to check that current tests are passed and make suggestions where further automations would help
```
[the remark this answers is not in the document]
Comment (RobJohnsonHO) [reply docs/patterns/code-reviews.md:49]: Just needs a full stop on the end here.

File: docs/patterns/code-reviews.md [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+
+Does the code protect against misuse?
+: Protecting against security vulnerabilities is crucial, checking for things such as allowing code injection, not validating user inputs or including Personally Identifiable Information (PII) in error messages will help to catch problems early.
+
+Do the changes consider existing code?
+: Reusability is important because it enhanced efficiency and minimises errors and leads to faster development and makes code easier to maintain. Frameworks, tools and standards within a team can also help a developer reuse functions they provide, make sure to check that code can't make use of these sources. Pay particular attention to this if the code may have been generated by AI which is not aware of shared functions/services etc.
+
+Is the code documented
+: Documentation ensures that teams can more easily maintain and developer a codebase further. Checking that you're happy with the level of documentation, especially if you were to work on this code in the future.
```
Comment (RobJohnsonHO) [inline docs/patterns/code-reviews.md:64]: maintain and develop a ...

**Review by RobJohnsonHO — COMMENTED** [review]
Just some minor grammatical things to update :)

**Review by RobJohnsonHO — APPROVED** [review]
LGTM
